File: built-in plugins dsp/mb-spring-reverb.cpp
/**
 * MB Spring Reverb
 * Category : effect
 * Type     : reverb
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic spring reverb emulation with drip and splash
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SPRING_REVERB_H
#define MB_SPRING_REVERB_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSpringReverb : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-spring-reverb";
    static constexpr const char* PLUGIN_NAME    = "MB Spring Reverb";
    static constexpr const char* PLUGIN_TYPE    = "reverb";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float mix = 0.3f;  // range [0, 1]
    float decay = 2f;  // range [0.5, 6]
    float tension = 0.5f;  // range [0, 1]
    float drip = 0.3f;  // range [0, 1]
    float diffusion = 0.5f;  // range [0, 1]
    float low_cut = 100f;  // range [20, 500]
    float high_cut = 8000f;  // range [1000, 20000]
    };

    MbSpringReverb() = default;
    ~MbSpringReverb() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.mix = std::clamp(params.mix, 0f, 1f);
        params.decay = std::clamp(params.decay, 0.5f, 6f);
        params.tension = std::clamp(params.tension, 0f, 1f);
        params.drip = std::clamp(params.drip, 0f, 1f);
        params.diffusion = std::clamp(params.diffusion, 0f, 1f);
        params.low_cut = std::clamp(params.low_cut, 20f, 500f);
        params.high_cut = std::clamp(params.high_cut, 1000f, 20000f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Spring Reverb
        return input;
    }
};

#endif // MB_SPRING_REVERB_H

File: built-in plugins dsp/mb-spring-reverb.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSpringReverbPlugin: PluginDefinition = {
    id: 'mb-spring-reverb',
    slug: 'mb-spring-reverb',
    name: 'MB Spring Reverb',
    category: 'effect',
    type: 'reverb',
    version: '1.0.0',
    description: 'Classic spring reverb emulation with drip and splash',
    author: 'Max Booster',
    parameters: [
      { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true },
      { id: 'decay', name: 'Decay', type: 'float', defaultValue: 2, minValue: 0.5, maxValue: 6, unit: 's', automatable: true },
      { id: 'tension', name: 'Spring Tension', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'drip', name: 'Drip', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true },
      { id: 'diffusion', name: 'Diffusion', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'low_cut', name: 'Low Cut', type: 'float', defaultValue: 100, minValue: 20, maxValue: 500, unit: 'Hz', automatable: true },
      { id: 'high_cut', name: 'High Cut', type: 'float', defaultValue: 8000, minValue: 1000, maxValue: 20000, unit: 'Hz', automatable: true },
    ],
    defaultPreset: { mix: 0.3, decay: 2, tension: 0.5, drip: 0.3, diffusion: 0.5, low_cut: 100, high_cut: 8000 },
  };

export default MbSpringReverbPlugin;
